labrad/fasttypes.c

```diff
@@ -0,0 +1,2364 @@
+#define UINT32 4
+#define FLOAT64 5
+#define COMPLEX128 6
+#define FLOAT64_SU 7
+#define COMPLEX128_SU 8
+#define STRING 9
+#define CLUSTER 10
+#define LIST 11
+#define NDARRAY 12
+#define NDARRAY_SU 13
+#define TIME 14
+#define ERROR 15
+
+// COBJ is an all-encompassing structure containing a wide enough variety of elements to represent any of the above types
+typedef struct cobj COBJ;
+
+struct cobj {
+    int type; // type of this cobj
+    int subtype; // if NDARRAY(_SU), type of each element in array, if ERROR, specific type of ERROR
+    int istype; // FALSE unless structure contains only type information
+    int i; // good for BOOL, UINT31, INT32
+    unsigned int n; // good for UINT32, length cobj_arr, length string, length dim_arr
+    double x, y; // FLOAT64 and COMPLEX128 data
+    char *str; // STRING (not owned), unit name (owned), and ERROR data (not owned)
+    COBJ **cobj_arr; // sub cobjs contained in CLUSTER or LIST (owned)
+    int *dim_arr; // dimensions of NDARRAY(_SU)
+    void *arr; // data of NDARRAY(_SU)
+    int isarrowned; // FALSE unless arr has be recreated to be a different type, usually arr is just a pointer to data in the pyobject.
+    long long int secs, usecs; // TIME data
+};
+
+// the following is a simplification of the real numpy PyArrayObject, use with caution
+typedef struct PyArrayObject {
+    PyObject_HEAD
+    char *data;
+    int nd;
+    int *dimensions;
+    int *strides;
+    PyObject *base;
+    void *descr;
+    int flags;
+    PyObject *weakreflist;
+} PyArrayObject;
+
+// global variables, initialized by first function call (which will take much longer than subsequent calls)
+int ft_gv_initialized = FALSE;
+int ft_gv_system_big_endian = FALSE;
+int ft_gv_send_big_endian = TRUE;
+PyObject *ft_gv_0 = NULL;
+PyObject *ft_gv_1 = NULL;
+PyObject *ft_gv_2 = NULL;
+PyObject *ft_gv_int_num_bits = NULL;
+PyObject *ft_gv_int_num_bits_minus_1 = NULL;
+PyObject *ft_gv_INT_MIN = NULL;
+PyObject *ft_gv_INT_MAX = NULL;
+PyObject *ft_gv_UINT_MAX = NULL;
+PyObject *ft_gv_numpy = NULL;
+PyTypeObject *ft_gv_numpy_ndarray = NULL;
+PyTypeObject *ft_gv_numpy_bool_ = NULL;
+PyTypeObject *ft_gv_numpy_int32 = NULL;
+PyTypeObject *ft_gv_numpy_uint32 = NULL;
+PyTypeObject *ft_gv_numpy_float64 = NULL;
+PyTypeObject *ft_gv_numpy_complex128 = NULL;
+PyObject *ft_gv_labrad_types = NULL;
+PyObject *ft_gv_labrad_units = NULL;
+PyTypeObject *ft_gv_labrad_units_WithUnit = NULL;
+PyTypeObject *ft_gv_labrad_units_Value = NULL;
+PyTypeObject *ft_gv_labrad_units_Complex = NULL;
+PyTypeObject *ft_gv_labrad_units_ValueArray = NULL;
+
+// C equivalent of labrad.types.flatten, called in similar manner, although type tags must be strings
+PyObject *ft_flatten(PyObject *self, PyObject *args, PyObject *keywds);
+
+// functions to create an internal C representation (COBJ) of supported Python objects
+COBJ *ft_create_cobj_from_pyobj(PyObject *obj);
+    COBJ *ft_allocate_memory_for_cobj();
+    COBJ *ft_create_empty_cobj();
```

Comment:
These should be declared explicitly (void).  () doesn't catch errors at call sites that pass unwanted arguments for compatibility with pre-c89 that didn't have prototypes.  (I will push a commit with this and other changes to fix compiler warnings)